spec/services/users/update_spec.rb

```diff
@@ -64,6 +64,43 @@ def sidekiq_assert_resave_article_worker(user, &block)
     expect(service.errors_as_sentence).to eq "filename too long - the max is 250 characters."
   end
 
+  context "when changing username" do
+    let(:new_username) { "#{user.username}_changed" }
+
+    it "sets old_username and old_old_username when username was changed" do
+      old_username = user.username
+      old_old_username = user.old_username
+      described_class.call(user, user: { username: new_username })
+      user.reload
+      expect(user.username).to eq(new_username)
+      expect(user.old_username).to eq(old_username)
+      expect(user.old_old_username).to eq(old_old_username)
+    end
+
+    it "changes user's articles path" do
+      article = create(:article, user: user)
+      old_path = article.path
+      sidekiq_perform_enqueued_jobs do
+        described_class.call(user, user: { username: new_username })
+      end
+      article.reload
+      expect(article.path).not_to eq(old_path)
+      expect(article.path).to eq("/#{new_username}/#{article.slug}")
+    end
+
+    # testing against gsub'ing username
+    it "sets the correct article path when its slug contains username" do
```

Comment:
I added a test against this situation described in the
https://github.com/forem/forem/issues/15424#issuecomment-975676049:
> Refine the logic so it's replacing the path's slug (not just the username). (Consider if the original user name is in the path twice) 

Actually, the previous code was working correctly in this case because after setting an incorrect `article.path` the `path` was rewritten in an `Article#set_caches` callback, which made setting `article.path` useless :grimacing: